### DISCUSSION TITLE
Sixel encoder/decoder

### DISCUSSION THREAD

**Discussion by Nutzzz** [SixLabors/ImageSharp#2796] [Ideas]
Especially now that Windows will be supporting Sixel in ConPty/Windows Terminal (Note Windows Terminal Canary supports it today, and it should be in mainline soon, probably in the next couple of weeks), perhaps an encoder/decoder could be added to ImageSharp?  I'd really love it here, otherwise I'd have to either do a PInvoke of libsixel or switch to Magick.NET.  Thanks!

**Comment by JimBobSquarePants** [comment]
That’s a pretty obscure format so we wouldn’t dedicate any time to supporting it any time soon. Perhaps via community contribution though.

**Reply by Nutzzz** [reply]
Sure, but that was my point; its obscurity has been partly due to the relative lack of support, which is changing.

I made a quick C# proof of concept that sort of works, but it currently only uses one color per 6-pixel high block (so it looks a bit like ANSI art).  I'll try to continue developing it.

**Reply by JimBobSquarePants** [reply]
Ah that’s excellent! Are you using libsixel as reference? 

https://github.com/saitoha/libsixel

The code there looks fairly readable.

**Reply by Nutzzz** [reply]
I used [this](https://github.com/hiromi-mi/sixel-sharp) "experimental" library in C# that turned out to be completely broken.  My proof of concept was, for fun, attempting to fix it based only on the sixel Wikipedia article.  As far as using libsixel (or chafa, perhaps) as reference... to the extent I'm good at any language, standard C ain't it.

**Comment by trackd** [comment]
Hey!
Me and ShaunLawrie made a Sixel PowerShell Module using Sixlabors lib.
[trackd/Sixel](https://github.com/trackd/Sixel) 
It works quite well and supports 256 colors, transparent etc.
(Max in WT)

Also Shaun has a PR open in SpectreConsole to add the sixel code there as well :)

So should be fairly accessible soon.

**Reply by JimBobSquarePants** [reply]
Very interesting! It looks like a lot of the code could be borrowed to write an instance of `IImageEncoder`. I couldn't find decoding code though? How does that work.

**Reply by Nutzzz** [reply]
> Also Shaun has a PR open in SpectreConsole to add the sixel code there as well :)

Note that [Sixel support](https://github.com/gui-cs/Terminal.Gui/pull/3734) was recently added to the V2 prerelease of the other major C# TUI framework, Terminal.Gui.

**Reply by trackd** [reply]
> Very interesting! It looks like a lot of the code could be borrowed to write an instance of `IImageEncoder`. I couldn't find decoding code though? How does that work.

correct, i never made a decoder. I didn't really have a need for one.

[SixPix.NET](https://github.com/teramako/SixPix.NET/) does have a decoder, but i havn't tried it.

> > Also Shaun has a PR open in SpectreConsole to add the sixel code there as well :)
> 
> Note that [Sixel support](https://github.com/gui-cs/Terminal.Gui/pull/3734) was recently added to the V2 prerelease of the other major C# TUI framework, Terminal.Gui.

Yea i saw that a few days ago, looks interesting. havn't had a chance to test it.

**Reply by JimBobSquarePants** [reply]
Looking at SixPix.NET it looks like decoding is trivial however I'm struggling to find any specifications for the format and specifically:

- How alpha support works.
- Can Sixel support multi-frame images?

I'm absolutely happy to work with someone to add official encoders and decoders to the library for V4.

**Reply by trackd** [reply]
> Looking at SixPix.NET it looks like decoding is trivial however I'm struggling to find any specifications for the format and specifically:
> 
>     * How alpha support works.

It needs to be enabled in the start parameters `1` (`"{ESC}P0;1q"`) to enable transparent mode.
and then you reference color 0 => `"#0;2;0;0;0";` with sixel `?`

so in my code it's this line that adds the transparent sixel.
`Append(color != 0 ? e : Constants.SIXELEMPTY)`

[graphicrenditions.md](https://github.com/hackerb9/vt340test/blob/main/docs/standards/graphicrenditions.md#iso-8613-6-direct-color-selection) this repo has alot of information and the original docs.

[wt sixel parser](https://github.com/microsoft/terminal/blob/main/src/terminal/adapter/SixelParser.cpp) is pretty readable and with alot of comments.

>     * Can Sixel support multi-frame images?

sort of, but it's pretty limited.. WT supports 6 pages.
https://github.com/microsoft/terminal/pull/17421#issuecomment-2177165191
the first example there is one of the few i've ever seen.
Most animations using sixel use some sort of script/code to orchestrate the timing & drawing
